capstone/src/test/java/com/google/step/RestaurantTest.java

```diff
@@ -0,0 +1,90 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.step;
+
+import java.util.ArrayList;
+import java.util.List;
+import com.google.appengine.api.datastore.Entity;
+import com.google.step.data.Restaurant;
+import com.google.appengine.api.datastore.GeoPt;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+import com.google.appengine.tools.development.testing.LocalUserServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalDatastoreServiceTestConfig;
+
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.After;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** */
+@RunWith(JUnit4.class)
+public final class RestaurantTest {
+  private final LocalServiceTestHelper helper =
+      new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());
+
+  @Before
+  public void setUp() {
+    helper.setUp();
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+/**
+ * Test that the factory method returns null when
+ * given the wrong entity type.
+ */
+  @Test
+  public void incorrectEntityKind() {
+    Entity wrongKind = new Entity("RestaurantUser");
+    Restaurant actual = Restaurant.fromEntity(wrongKind);
```

Comment:
While it's been a while since I've coded in Java, I skimmed the effective java section on exceptions, and it seems the recommended pattern is to provide a state checker, and then throw exceptions for code errors/bugs.

So consider the following API instead of returning null (it might make the code easier to read/understand):
if (!Restaurant.isRestaurantEntity(entity)) {
  // Handle this issue
}
Restaurant restaurant = Restaurant.fromEntity(entity); // Throws InvalidArgumentException if the entity has the wrong type.

Like I said it's been a while since I've written Java, and I could be totally off base. If this feels like a strange design please feel free to say so.

Reply:
Thank you, this makes much more sense. Updated and will keep in mind for later!